File: _6.7_hanoi_tower.py
'''
Hanoi Tower

# Initial State:
Disks are stacked on the Source rod from largest (bottom) to smallest (top).

# Final State:
Disks are moved to the Target rod, maintaining the same order.

# Constraints:
1. Move one disk at a time.
2. Place the disk on top of another stack or an empty rod.
3. Do not place a larger disk on top of a smaller one.
'''

# Solve the Hanoi Tower problem.
def hanoi(num_disks, source, auxiliary, target):
    if num_disks == 1:
        # Move 1 disk from source to target
        target.append(source.pop())
    else:
        # Move n-1 disks from source to auxiliary
        hanoi(num_disks - 1, source, target, auxiliary)
        # Move 1 disk from source to target
        target.append(source.pop())
        # Move n-1 disks from auxiliary to target
        hanoi(num_disks - 1, auxiliary, source, target)

# Print the current state of the rods.
def print_rods(source, auxiliary, target):
    print("Source Rod:", source)
    print("Auxiliary Rod:", auxiliary)
    print("Target Rod:", target)
    print("-" * 30)

# Get the number of disks from the user.
def get_num_disks():
    while True:
        try:
            num_disks = int(input("Enter the number of disks: "))
            if num_disks > 0:
                return num_disks
            else:
                print("Please enter a positive integer.")
        except ValueError:
            print("Invalid input. Please enter an integer.")

# Initialize the rods with disks.
def initialize_rods(num_disks):
    source_rod = list(range(num_disks, 0, -1))  # list from num_disks to 1
    auxiliary_rod = []
    target_rod = []
    return source_rod, auxiliary_rod, target_rod

# Get the number of disks
num_disks = get_num_disks()

# Initialize the rods
source_rod, auxiliary_rod, target_rod = initialize_rods(num_disks)

# Print initial state of the rods
print("Initial state:")
print_rods(source_rod, auxiliary_rod, target_rod)

# Solve the Hanoi Tower problem
hanoi(num_disks, source_rod, auxiliary_rod, target_rod)

# Print final state of the rods
print("Final state:")
print_rods(source_rod, auxiliary_rod, target_rod)

File: _6.7_hanoi-tower.js
/*
Hanoi Tower

* Initial State:
Disks are stacked on the Source rod from largest (bottom) to smallest (top).

* Final State:
Disks are moved to the Target rod, maintaining the same order.

* Constraints:
1. Move one disk at a time.
2. Place the disk on top of another stack or an empty rod.
3. Do not place a larger disk on top of a smaller one.
*/

const readline = require('readline');

// Solve the Hanoi Tower problem.
function hanoi(numDisks, source, auxiliary, target) {
    if (numDisks === 1) {
        // Move 1 disk from source to target
        target.push(source.pop());
    } else {
        // Move n-1 disks from source to auxiliary
        hanoi(numDisks - 1, source, target, auxiliary);
        // Move 1 disk from source to target
        target.push(source.pop());
        // Move n-1 disks from auxiliary to target
        hanoi(numDisks - 1, auxiliary, source, target);
    }
}

// Print the current state of the rods.
function printRods(source, auxiliary, target) {
    console.log("Source Rod:", source);
    console.log("Auxiliary Rod:", auxiliary);
    console.log("Target Rod:", target);
    console.log("-".repeat(30));
}

// Get the number of disks from the user.
function getNumDisks() {
    const rl = readline.createInterface({
        input: process.stdin,
        output: process.stdout
    });

    return new Promise((resolve) => {
        rl.question("Enter the number of disks: ", (answer) => {
            const numDisks = parseInt(answer, 10);
            if (numDisks > 0 && numDisks <= 20) {
                rl.close();
                resolve(numDisks);
            } else {
                console.log("Please enter a positive integer.");
                rl.close();
                resolve(getNumDisks());
            }
        });
    });
}

// Initialize the rods with disks.
function initializeRods(numDisks) {
    let sourceRod = Array.from({ length: numDisks }, (v, i) => numDisks - i);  // array from numDisks to 1
    let auxiliaryRod = [];
    let targetRod = [];
    return { sourceRod, auxiliaryRod, targetRod };
}

// Main function to run the Hanoi Tower solution
async function main() {
    // Get the number of disks
    let numDisks = await getNumDisks();

    // Initialize the rods
    let rods = initializeRods(numDisks);

    // Print initial state of the rods
    console.log("Initial state:");
    printRods(rods.sourceRod, rods.auxiliaryRod, rods.targetRod);

    // Solve the Hanoi Tower problem
    hanoi(numDisks, rods.sourceRod, rods.auxiliaryRod, rods.targetRod);

    // Print final state of the rods
    console.log("Final state:");
    printRods(rods.sourceRod, rods.auxiliaryRod, rods.targetRod);
}

main();
